File: pipelines/atlas_pipeline/hazard_graph.py
"""Publish source-linked drainage and conditional exposure relationships."""
import gzip
import hashlib
import json
import shutil

from jsonschema import Draft7Validator
from shapely.geometry import shape

from processing.hazard_graph.engine import validate_graph

from .contracts import ROOT
from .exposure import read_dataset
from .exposure_contracts import verify_exposure

BASE = '/data/nepal-hazard-graph/1.0.0/'


def encoded(value):
    return (json.dumps(value, sort_keys=True, separators=(',', ':'), allow_nan=False) + '\n').encode()


def digest(raw):
    return hashlib.sha256(raw).hexdigest()


def derive():
    inputs, nodes, edges, rivers = [], [], [], {}

    def reference(dataset, version, metadata, artifact_hash):
        raw = (ROOT / f'data/releases/{dataset}/{version}/manifest.json').read_bytes()
        inputs.append({'dataset_id': dataset, 'version': version, 'path': f'/data/{dataset}/{version}/manifest.json', 'sha256': digest(raw), 'artifact_sha256': artifact_hash,
                       'source': metadata.get('source', 'Atlas exposure overlay'), 'license': metadata.get('license', 'Underlying source licences retained in exposure manifest'), 'date': metadata.get('observation_date')})
        return len(inputs) - 1

    for dataset in ['nepal-rivers-primary', 'nepal-rivers-headwaters']:
        data = read_dataset(dataset)
        ref = reference(dataset, '1.0.0', data['metadata'], data['metadata']['artifact']['sha256'])
        for f in data['collection']['features']:
            p = f['properties']
            point = shape(f['geometry']).representative_point()
            node = {'id': 'river:' + p['source_id'], 'type': 'river', 'label': 'HYRIV ' + p['source_id'], 'input': ref, 'record_id': str(f['id']), 'coordinates': [point.x, point.y], 'boundary_next_id': p['downstream_id'] if not p['downstream_in_release'] else None}
            nodes.append(node)
            rivers[p['source_id']] = (node, p)
    for source, (node, p) in sorted(rivers.items()):
        target = p['downstream_id']
        if p['downstream_in_release'] != (target in rivers):
            raise ValueError('Inconsistent retained river topology')
        if target in rivers:
            edges.append({'id': f'next:{source}:{target}', 'from': node['id'], 'to': 'river:' + target, 'type': 'river_downstream', 'evidence': 'derived', 'confidence': None, 'input': node['input'], 'record_id': node['record_id'], 'method': 'HYRIV_ID/NEXT_DOWN source pointer', 'assumptions': [], 'limitations': ['Connectivity only; not measured flow, hazard propagation or travel time.']})
    for width in [250, 1000]:
        dataset = f'exposure-trace-40669746-{width}m'
        folder = ROOT / f'data/releases/{dataset}/1.0.0'
        result = verify_exposure(folder)
        ref = reference(dataset, '1.0.0', result, result['artifacts']['spatial']['sha256'])
        spatial = json.loads(gzip.decompress((folder / 'spatial.geojson.gz').read_bytes()))
        centre = shape(spatial['features'][0]['geometry']).representative_point()
        for prefix, type in [('scenario', 'scenario'), ('exposure', 'exposure')]:
            nodes.append({'id': prefix + ':' + dataset, 'type': type, 'label': f'Hypothetical {width} m corridor {prefix}', 'input': ref, 'record_id': 'request.json' if prefix == 'scenario' else result['result_id'], 'coordinates': [centre.x, centre.y], 'boundary_next_id': None})
        assumption = ['Conditioned on the recorded hypothetical corridor, not an observed hazard footprint.']
        edges.append({'id': 'scenario:' + dataset, 'from': 'scenario:' + dataset, 'to': 'exposure:' + dataset, 'type': 'scenario_exposure', 'evidence': 'modelled', 'confidence': None, 'input': ref, 'record_id': result['result_id'], 'method': result['method'], 'assumptions': assumption, 'limitations': ['Spatial exposure estimate, not vulnerability, damage or risk.']})
        positions = {f['properties']['asset_id']: shape(f['geometry']).representative_point() for f in spatial['features'] if f['properties']['kind'] == 'asset'}
        for asset in result['assets']:
            id = 'asset:' + dataset + ':' + asset['id']
            point = positions[asset['id']]
            nodes.append({'id': id, 'type': 'infrastructure', 'label': asset['name'] or asset['id'], 'input': ref, 'record_id': asset['id'], 'coordinates': [point.x, point.y], 'boundary_next_id': None})
            edges.append({'id': 'intersects:' + id, 'from': 'exposure:' + dataset, 'to': id, 'type': 'footprint_intersection', 'evidence': 'modelled', 'confidence': None, 'input': ref, 'record_id': asset['id'], 'method': result['method'], 'assumptions': assumption, 'limitations': ['Mapped asset intersects the assumed footprint; not confirmed impact.']})
    return validate_graph({'schema_version': '1.0.0', 'kind': 'hazard-graph', 'version': '1.0.0', 'method': 'evidence-relationships/1.0.0', 'inputs': inputs, 'nodes': sorted(nodes, key=lambda n: n['id']), 'edges': sorted(edges, key=lambda e: e['id']), 'limitations': ['No glacier-to-lake, lake outlet, hazard-to-exposure or landslide blockage links are asserted without supporting evidence.', 'Missing edges mean unknown relationships, not geographic absence. No causal or transitive hazard inference is performed.', 'Confidence is UNKNOWN where the sources provide no relationship accuracy. Asset nodes are scoped to their exposure result; matching OSM IDs do not create extra relationships.', 'Node coordinates are representative map positions, not proof of connectivity. River pointers leaving this release remain explicit coverage exits.']})


def verify_hazard_graph(path, public=None):
    manifest = json.loads((path / 'manifest.json').read_text())
    raw = (path / 'graph.json.gz').read_bytes()
    if len(raw) > 2_097_152 or manifest['artifact']['sha256'] != digest(raw) or manifest['artifact']['byte_size'] != len(raw):
        raise ValueError('Graph checksum/size mismatch')
    decoded = gzip.decompress(raw)
    if len(decoded) > 16_777_216 or len(decoded) != manifest['artifact']['decoded_byte_size']:
        raise ValueError('Graph decoded size mismatch')
    graph = json.loads(decoded)
    Draft7Validator(json.loads((ROOT / 'schemas/hazard-graph.schema.json').read_text())).validate(graph)
    validate_graph(graph)
    if graph != derive():
        raise ValueError('Graph differs from pinned source records')
    if public and any((path / name).read_bytes() != (public / name).read_bytes() for name in ['manifest.json', 'graph.json.gz']):
        raise ValueError('Public graph differs')
    return graph


def build():
    release, public = ROOT / ('data/releases' + BASE[5:]), ROOT / ('apps/web/public' + BASE)
    if release.exists():
        verify_hazard_graph(release, public)
        return
    graph = derive()
    decoded = encoded(graph)
    raw = gzip.compress(decoded, mtime=0)
    manifest = {'kind': 'hazard-graph-release', 'version': '1.0.0', 'artifact': {'path': BASE + 'graph.json.gz', 'sha256': digest(raw), 'byte_size': len(raw), 'decoded_byte_size': len(decoded)}}
    stage = ROOT / 'data/staging/hazard-graph'
    stage.mkdir(parents=True, exist_ok=True)
    (stage / 'graph.json.gz').write_bytes(raw)
    (stage / 'manifest.json').write_bytes(encoded(manifest))
    verify_hazard_graph(stage)
    release.parent.mkdir(parents=True, exist_ok=True)
    shutil.copytree(stage, release)
    shutil.copytree(stage, public)
    print(f"Published {len(graph['nodes'])} nodes and {len(graph['edges'])} edges")


if __name__ == '__main__':
    build()

File: packages/contracts/hazard-graph.ts
import Ajv from 'ajv';
import schema from '../../schemas/hazard-graph.schema.json';
export type GraphNodeType = 'glacier' | 'lake' | 'river' | 'hazard' | 'infrastructure' | 'population_area' | 'event' | 'scenario' | 'exposure';
export type GraphEvidence = 'observed' | 'derived' | 'inferred' | 'modelled';
export type Relation = 'glacier_lake' | 'lake_river' | 'river_downstream' | 'hazard_exposure' | 'landslide_blockage' | 'scenario_exposure' | 'footprint_intersection';
export interface GraphNode { id: string; type: GraphNodeType; label: string; input: number; record_id: string; coordinates: [number, number] | null; boundary_next_id: string | null }
export interface GraphEdge { id: string; from: string; to: string; type: Relation; evidence: GraphEvidence; confidence: 'low' | 'medium' | 'high' | null; input: number; record_id: string; method: string; assumptions: string[]; limitations: string[] }
export interface HazardGraph { schema_version: '1.0.0'; kind: 'hazard-graph'; version: '1.0.0'; method: 'evidence-relationships/1.0.0'; nodes: GraphNode[]; edges: GraphEdge[]; inputs: { dataset_id: string; version: string; path: string; sha256: string; artifact_sha256: string; source: string; license: string; date: string | null }[]; limitations: string[] }
export const RELATIONS: Record<Relation, { from: GraphNodeType[]; to: GraphNodeType[]; meaning: string }> = {
  glacier_lake: { from: ['glacier'], to: ['lake'], meaning: 'Documented glacier–lake relationship; proximity alone is insufficient.' },
  lake_river: { from: ['lake'], to: ['river'], meaning: 'Documented lake outlet connection, not a nearest-river assignment.' },
  river_downstream: { from: ['river'], to: ['river'], meaning: 'Source NEXT_DOWN drainage pointer; not hazard propagation.' },
  hazard_exposure: { from: ['hazard'], to: ['exposure', 'population_area', 'infrastructure'], meaning: 'Evidence-backed footprint overlay; not damage or risk.' },
  landslide_blockage: { from: ['event', 'hazard'], to: ['river'], meaning: 'Documented blockage evidence or explicit model/inference; not proximity.' },
  scenario_exposure: { from: ['scenario'], to: ['exposure'], meaning: 'Exposure estimate conditional on a hypothetical corridor.' },
  footprint_intersection: { from: ['exposure'], to: ['infrastructure', 'population_area'], meaning: 'Mapped record intersects an assumed footprint; not confirmed impact.' },
};
const ajv = new Ajv({ allErrors: true, strict: true });
const validate = ajv.compile<HazardGraph>(schema);
export function parseHazardGraph(input: unknown): HazardGraph {
  if (!validate(input)) throw new Error(`Invalid hazard graph: ${ajv.errorsText(validate.errors)}`);
  const nodes = new Map(input.nodes.map(n => [n.id, n]));
  if (nodes.size !== input.nodes.length) throw new Error('Duplicate graph nodes');
  const ids = new Set<string>(), links = new Set<string>();
  for (const n of input.nodes) {
    if (!input.inputs[n.input] || (n.coordinates && !(n.coordinates[0] >= 79 && n.coordinates[0] <= 89 && n.coordinates[1] >= 25 && n.coordinates[1] <= 32))) throw new Error('Invalid node source or CRS84 position');
  }
  for (const e of input.edges) {
    const a = nodes.get(e.from), b = nodes.get(e.to), type = RELATIONS[e.type];
    const key = JSON.stringify([e.from, e.to, e.type]);
    if (!a || !b || a.id === b.id || ids.has(e.id) || links.has(key)) throw new Error('Missing node or duplicate/self edge');
    if (!type.from.includes(a.type) || !type.to.includes(b.type) || !input.inputs[e.input]) throw new Error('Invalid edge type or evidence source');
    if (['inferred', 'modelled'].includes(e.evidence) && !e.assumptions.length) throw new Error('Inference/model output requires explicit assumptions');
    if (e.type === 'river_downstream' && e.evidence !== 'derived') throw new Error('Source drainage topology is derived');
    if (['scenario_exposure', 'footprint_intersection'].includes(e.type) && e.evidence !== 'modelled') throw new Error('Hypothetical relationships must remain modelled');
    ids.add(e.id); links.add(key);
  }
  return input;
}
export function traverseGraph(graph: HazardGraph, start: string, limit = 50) {
  if (!Number.isInteger(limit) || limit < 1 || limit > 500 || !graph.nodes.some(n => n.id === start)) throw new Error('Invalid graph traversal request');
  const adjacency = new Map<string, GraphEdge[]>();
  for (const e of [...graph.edges].sort((a, b) => a.id < b.id ? -1 : a.id > b.id ? 1 : 0)) {
    const outgoing = adjacency.get(e.from);
    if (outgoing) outgoing.push(e); else adjacency.set(e.from, [e]);
  }
  const queue = [start], visited = new Set<string>(), edges = new Set<string>();
  let cursor = 0;
  while (cursor < queue.length && visited.size < limit) {
    const id = queue[cursor++]; if (visited.has(id)) continue;
    visited.add(id);
    for (const edge of adjacency.get(id) ?? []) { edges.add(edge.id); if (!visited.has(edge.to)) queue.push(edge.to); }
  }
  return { nodes: [...visited].sort(), edges: [...edges].sort(), truncated: queue.slice(cursor).some(id => !visited.has(id)) };
}
